File: PRACTICA1/RUBY/Deepspace/NumericDamage.rb

#encoding:utf-8

module Deepspace

require_relative "NumericDamageToUI"
        
        # @author Lucía Salamanca López, Celia Arias Martínez
        
class NumericDamage < Damage
    def initialize(n,s)  #n numero de weapons, s numero de shield
        super(s)
        @nWeapons = n
    end

    def copy
        NumericDamage.new(@nWeapons, @nShields)
    end

    def getUIversion
        NumericDamageToUI.new(self)
    end

    def adjust(w,s)  #w es array de weapons, s array de shieldbooster
        shields = super(s)
        if (@nWeapons > w.length)
            return NumericDamage.new(w.length,shields)
        else
            return NumericDamage.new(@nWeapons,shields)
        end
    end

    def discardWeapon(w)
        if (@nWeapons > 0)
            @nWeapons = @nWeapons-1
        end
    end

    def hasNoEffect
        return ((@nWeapons == 0) && (super))
    end

    def nWeapons
        @nWeapons
    end

    def to_s
        super + "#{@nWeapons} armas."  
    end
end

end

File: PRACTICA1/JAVA/deepspace/src/deepspace/NumericDamage.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author Celia Arias Martínez, Lucía Salamanca López
 */
public class NumericDamage extends Damage{
    private int nWeapons;
    
    public NumericDamage(int w, int s ){
        super(s);
        nWeapons = w;
    }
    
    public NumericDamage copy(){
        return new NumericDamage(nWeapons, getNShields());
    }
    
    @Override
    public NumericDamage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s){
        int shields = super.adjust(s);
        if (nWeapons > w.size())
            return new NumericDamage(w.size(),shields);
        else 
            return new NumericDamage(nWeapons,shields);
    }
    
    @Override
    public void discardWeapon(Weapon w){
        if (nWeapons>0)
            nWeapons --;
    }
    
    @Override
    public boolean hasNoEffect(){
        return ((nWeapons == 0) && super.hasNoEffect());
    }
    
    public int getNWeapons(){
        return nWeapons;
    }
    
    @Override
    public String toString(){
        String mensaje = super.toString()+"y "+nWeapons+" armas.";
        return mensaje;
    }
    
    NumericDamageToUI getUIversion(){
        return new NumericDamageToUI(this);
    }
}
